### ISSUE TITLE
Optional inclusion of `assert` for `reportUnnecessaryIsInstance` and `reportUnnecessaryComparison`

### ISSUE THREAD

**Issue by tpgillam** [microsoft/pyright#9046]
We have been working in a moderately sized codebase to add typechecking. It's a long process, and as intermediate steps we've had to add a lot of statements to verify intent, like `assert thing.x is not None` and `assert isinstance(thing, Moo)`.

Whilst I fully agree with not suggesting removal of unnecessary comparisons in `assert` statements by default (e.g. as described here: https://github.com/microsoft/pyright/issues/3437#issuecomment-1120144244 ), in our case we would like to temporarily enable it to allow us to quickly find and eliminate unnecessary statements.

I'm not familiar with the internals of pyright and whether this would be easy to support, but it would certainly save us some time. (Our other idea would be to use some sed / awk to rewrite all assertions as explicit `if` statements, then run the existing `reportUnnecessaryIsInstance` etc. to trim down, then transform the ones still left back).

I'm not sure whether it would be better to have an option that modified the behaviour of the existing rules, or to add new rules `reportUnnecessaryAssertIsInstance` `reportUnnecessaryAssertComparison` that is never enabled by default. I can see advantanges / disadvantages either way.

**Comment by erictraut** [comment]
I think this problem is pretty specific to your use case, so I'm reluctant to add two more diagnostic checks specifically for unnecessary isinstance and comparison expressions within asserts. I might consider this in the future if there's strong signal from other pyright users that this is something that's generally desired, but for now I'm going to reject this enhancement request.

Asserts generally don't do any harm, so you could simply leave them in place in your code. 

If you don't feel comfortable with that suggestion, here are a few other ideas:
1. Write a script that that temporarily modifies your code to search for `assert` statements and inserts the condition expression before the assert on the same line, then run pyright on the resulting code. It will then flag any unnecessary conditions. These should be on the same line as the original assert, so it will be easy to find and remove the unnecessary asserts in the original (unmodified) code base.
```python
# pyright: reportUnnecessaryIsInstance=warning
x = 1

# Original code
assert isinstance(x, str)

# After script runs
isinstance(x, str); assert isinstance(x, str)
```

2. Check out the pyright sources and build your own custom variant that modifies this specific behavior. The change needs to be made [here](https://github.com/microsoft/pyright/blob/32a5ac5d2a3e063960851148d4dba0bba5197348/packages/pyright-internal/src/analyzer/parseTreeUtils.ts#L1541). Simply comment out most of the code in the `isWithinAssertExpression` function so it always returns `false`. For pyright build instruction, refer to [this documentation](https://microsoft.github.io/pyright/#/build-debug).

**Comment by tpgillam** [comment]
Thank you for the consideration and useful suggestions!
